### DISCUSSION TITLE
Get column names for row entries that have improper formatting/junk data?

### DISCUSSION THREAD

**Discussion by ILTCED** [JoshClose/CsvHelper#2049] [Q&A]
I would like to do the following:
1. Iterate over a csv file
2. When a row entry is discovered to be improperly formatted or has junk input, get the corresponding Column name and store in a list
3. Throw an error that returns that list to the user so they can correct their mistakes.

For example, if i had:
DateTime, Amount
03/14/2021, sdfd,
s%$3, 7

The program could see the Amount value in row 1 is bad, and grab the column name "Amount" and put into a list.
Likewise, in row 2 the DateTime row entry is detected and bad, and the column name "DateTime" is put into the list.
Then if that list contains values, an error is thrown to the user.

I don't have code for this, just wondering if it is possible and what tools can i use?
Currently I use Validate for each model variable, but that stops the program on the first error discovered.

**Comment by JoshClose** [comment]
Here is a way you can get type conversion errors and bad data errors. Keep in mind that some bad data could completely ruin reading the rest of the file.

```cs
void Main()
{
    var s = new StringBuilder();
    s.Append("Id,Name\r\n");
    s.Append("1,one\r\n");
    s.Append("2a,two\r\n");
    s.Append("3,three\r\n");
    s.Append("b4,four\r\n");
    s.Append("5,five\"foo\r\n");
    
    var records = new List<Foo>();
    var errors = new List<string>();
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.Context.RegisterClassMap<FooMap>();
        csv.Read();
        csv.ReadHeader();
        while (csv.Read())
        {
            try
            {
                records.Add(csv.GetRecord<Foo>());
            }
            catch (BadDataException ex)
            {
                errors.Add($"BadData: Line: {csv.Parser.RawRow}, Record: '{csv.Parser.RawRecord}'");
            }
            catch (TypeConverterException ex)
            {
                errors.Add($"TypeConversion: Line: {csv.Parser.RawRow} Field: '{ex.Text}', Record: '{csv.Parser.RawRecord}'");
            }
        }        
    }

    errors.Dump();
    records.Dump();
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
}

private class FooMap : ClassMap<Foo>
{
    public FooMap()
    {
        Map(m => m.Id);
        Map(m => m.Name);
    }
}
```
![image](https://user-images.githubusercontent.com/114971/194357995-dadd743a-2b62-423a-8f83-5bd010f76b07.png)

**Comment by ILTCED** [comment]
Thank you for the code example! I will need to play around with that.

Regarding the raw record, how would i get the particular row value that caused the error for that column?

Suppose:
DateTime, Id, Name
3/fdsd/2021, 4, Todd

So for the raw record of row one, I want the DateTime value, and not everything else

**Reply by JoshClose** [reply]
> Regarding the raw record, how would i get the particular row value that caused the error for that column?

`csv.Parser.RawRow`

> So for the raw record of row one, I want the DateTime value, and not everything else

In the next release, the field will be a part of `BadDataException` and that is what you want. With `TypeConverterException` it's already available and in the example.

**Reply by ILTCED** [reply]
thank you!
